libyul/ControlFlowSideEffectsCollector.cpp

```diff
@@ -0,0 +1,269 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/ControlFlowSideEffectsCollector.h>
+
+#include <libyul/optimiser/FunctionDefinitionCollector.h>
+
+#include <libyul/AST.h>
+#include <libyul/Dialect.h>
+
+#include <libsolutil/Common.h>
+#include <libsolutil/Algorithms.h>
+
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/algorithm/find_if.hpp>
+
+using namespace std;
+using namespace solidity::yul;
+
+
+ControlFlowBuilder::ControlFlowBuilder(Block const& _ast)
+{
+	for (auto const& statement: _ast.statements)
+		if (auto const* function = get_if<FunctionDefinition>(&statement))
+			(*this)(*function);
+}
+
+void ControlFlowBuilder::operator()(FunctionCall const& _functionCall)
+{
+	walkVector(_functionCall.arguments | ranges::views::reverse);
+	newConnectedNode();
+	m_currentNode->functionCall = _functionCall.functionName.name;
+}
+
+void ControlFlowBuilder::operator()(If const& _if)
+{
+	visit(*_if.condition);
+	ControlFlowNode* node = m_currentNode;
+	(*this)(_if.body);
```

Comment:
I would rather have expected
```suggestion
	newConnectedNode();
	(*this)(_if.body);
```
here...

Reply:
And similarly in the ``Switch``... but it might not matter as long as we don't start trying to store anything inside the nodes and every function call get's it's own node...

Reply:
We have to create a new node after the body, because whenever there is a function call, a new node is created and the function call is added. This means If there is a function call at the end of the body, the condition node should not be connected to that node. Creating a node before the body is optional, though, I think.

Reply:
Yes, there definitely needs a new one after the body.
And creating a new one before the body may indeed be optional right now, but we need to keep aware of it, if we ever want to use this graph for other purposes... if e.g. the graph should contain a list of assignments in each node, then for
```
if cond { x := 4 }
```
the assignment to ``x`` ends up in the same node as the condition... or every assignment would need to create a new node...

Reply:
But yeah, it's probably fine for the current purpose in any case.